gibbon-compiler/src/Gibbon/Passes/OptimizeADTLayout.hs

```diff
@@ -246,43 +249,116 @@ globallyOptimizeDataConLayout useGreedy prg@Prog{ddefs, fundefs, mainExp} = do
+              (\(f@FunDef{funName=fname}, (dcon, newSymDcon), constrs) pr@Prog{ddefs=dd, fundefs=fds, mainExp=mexp} -> do 
                       let maybeFd = M.lookup fname fds
-                      let fieldOrder = M.insert fname edgeOrder M.empty
                       let fd = case maybeFd of
                                                   Just x -> x
                                                   Nothing -> error "globallyOptimizeDataConLayout: expected a function definition!!" 
-                      let result = optimizeFunctionWRTDataCon dd fd dcon (fromVar newSymDcon) fieldOrder useGreedy
+                      let result = optimizeFunctionWRTDataConGlobal dd fd dcon (fromVar newSymDcon) (Left constrs) useGreedy
                       case result of
-                                     Nothing ->  pure pr 
+                                     Nothing -> pure pr 
                                      Just (ddefs', fundef', fieldorder) -> let fundefs' = M.delete fname fds
                                                                                fundefs'' = M.insert fname fundef' fundefs'
                                                                                p = Prog{ddefs = ddefs', fundefs = fundefs'', mainExp = mexp}
-                                                                               venv = progToVEnv p
-                                                                               pmap = generateProducerGraph p
-                                                                               prg' = genNewProducersAndRewriteProgram fname (fromVar newSymDcon) fieldorder venv pmap p
-                                                                             in  pure prg'
+                                                                               prg' = globallyChangeDataConstructorLayout dcon (fromVar newSymDcon) fieldorder p
+                                                                             in pure prg'
                 ) 
 
-  let lambda = (\(fList, dcon, fieldOrder) pr@Prog{ddefs=dd, fundefs=fds, mainExp=mexp} -> do
+  let solverMain = 
+              (\(fList, dcon, constrs) pr@Prog{ddefs=dd, fundefs=fds, mainExp=mexp} -> do
                                    newSymDcon  <- gensym (toVar dcon)
-                                   let vals = P.map (\f -> (f, (dcon, newSymDcon), fieldOrder)) fList 
-                                   P.foldrM lambda' pr vals  
-               )
-  
-
-  P.foldrM lambda prg funsToOptimizeTriple
+                                   let vals = P.map (\f -> (f, (dcon, newSymDcon), constrs)) fList 
+                                   P.foldrM solverOptimization pr vals
+              )
 
+  let greedyOptimization = 
+                  (\(f@FunDef{funName=fname}, (dcon, newSymDcon), edges) pr@Prog{ddefs=dd, fundefs=fds, mainExp=mexp} -> do 
+                          let maybeFd = M.lookup fname fds 
+                          let fd = case maybeFd of 
```

Comment:
You tend to write in ANF, I think :-) This makes code longer than it needs to be and introduces a bunch of identifiers, which spends the reader's mental energy.
```suggestion
                          let fd = case M.lookup fname fds of 
```